### ISSUE TITLE
Magento merged JavaScript throws errors

### ISSUE THREAD

**Issue by danemacmillan** [ebizmarts/magemonkey#131]
### Situation - How did you get there?
- Turn Magento's JavaScript merging on.
- Be unfortunate enough to have one of Magento's js/varien/*.js files concatenated directly above the `Ebizmarts_MageMonkey` JavaScript IIFE.
- Watch frontend crash and burn.

### Problem - What went wrong?
Magento makes heavy use of object literal syntax for defining objects. They take the following form:
```
Product.Config.prototype = {
}
```
The problem is that they *should* be ending each object literal with a semicolon, but they are not. This can be very problematic should certain conditions be met. In their own distinct files this is no problem. The problem arises when merging all these files together. A vanilla Magento install will not have the issue; it won't have the issue simply because it has been lucky enough, but that's it: the problem has always been there&mdash;just dormant, waiting.

Now, an IIFE is a technique to invoke a function expression. It is done so using `()` at the end of a function expression, which is defined within curly braces. Placing the `()` after the ending curly brace instructs JavaScript to try and invoke the body within the curly braces, to be simple.

What does this mean for you guys?

It means that should any of Magento's native JavaScript, which defines most of its functionality as object literals and does not end them with semicolons, immediately precede the JavaScript within your module, which is all defined as IIFEs, your IIFEs will actually cause Magento's object literals to attempt to be invoked, because it will look like this:

```js
Product.Config.prototype = {
}
/*!
 * Ebizmarts_MageMonkey
 * @copyright   Copyright Ebizmarts
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 *
 */
(function () {
})();
```

Let me take out the comment block and put this all together to be more clear:
```js
Product.Config.prototype = {
}(function () {
})();
```

Your IIFE is serving as an invocation instruction on the object literal preceding it. In other words, you went full Xzibit, dawg.

### Possible solution - How should it be?
It is because of situations like this where a vendor has no control over the environment that their JavaScript is executed that it is very much recommended to prefix all IIFEs, if the technique is being used, with a semicolon. This is also known as a *defensive semicolon*, should you want to research.

In other words, do this and the problem will be solved:
```js
Product.Config.prototype = {
}
/*!
 * Ebizmarts_MageMonkey
 * @copyright   Copyright Ebizmarts
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 *
 */
;(function () {
})();
```
Note the semicolon preceding the opening parenthesis.

That means the JavaScript interpreter will see the code like this and not choke:
```js
Product.Config.prototype = {
};(function () {
})();
```
